napari/_vispy/layers/base.py

```diff
@@ -195,6 +195,13 @@ def _on_matrix_change(self):
         translate = transform.translate[::-1]
         matrix = transform.linear_matrix[::-1, ::-1].T
 
+        if (
+            self._array_like
+            and self.layer._slice_input.ndisplay == 3
+            and self.layer.multiscale
+        ):
+            translate = translate + (self.layer.downsample_factors[-1] - 1) / 2
```

Comment:
Why do you subtract 1 from the last level's downsample factor? For a downsample factor of 1, that would be 0, which seems right, but for a downsample factor of 4, that would be 3, which doesn't seem right because we'll get an offset of 1.5 after the divide by 2.

Ideally, I'd prefer to handle this with the similar 2D array-like case below, which should work automatically due to the way [the tile2data transform is defined in slicing](https://github.com/napari/napari/blob/d02b8569270532cf9738b1bbc0ef90b926513b44/napari/layers/image/_slice.py#L219). But maybe I'm missing a detail here.

Reply:
> For a downsample factor of 1, we eventually get an offset of 0 whereas I'd expect 0.5

No, because we don't want to offset the "base" scale. We do want the origin of the highest resolution to be at the center of the first pixel, to match every other thing in napari. If you leave the `-1` out, you end up shifting only multiscale layers so that they have the corner of the first pixel at the origin.

I hope this makes sense, I'm still jetlagged xD

> Ideally, I'd prefer to handle this with the similar 2D array-like case below

I tried to use the `tile2data` transform, but things broke down in way I could not figure out, and I was afraid of messing up unrelated things. If you know how to attack that, please do! In principle, I agree that this should be unified. On the other hand, there is a confusion/mixing of "multiscale" and "tiling" in this whole part of the code that makes it hard to work with, so it would be great if we could separate them more first.

Reply:
Will try to come back to this today or later in the week.

Agreed that these code paths are confusing. Maybe some motivation for @kephale to bring in multi-scale virtual data soon?

Reply:
Thought some more on this and we shouldn't use `tile2data` here - that should only account for a partial image slice (e.g. in the 2D case) within the whole data, not for this extra offset which is for rendering.

Reply:
Just to inspire confidence, I'm going to flip-flop again. But this time it's OK because I have a diagram!

![napari-tile2data drawio](https://github.com/napari/napari/assets/2608297/4366da05-ae58-4709-ab59-b41644d638df)

I'm going to claim that the `tile2data` transform captures part of information we care about here. The critical thing (which I missed and you alluded to) is that the `data` in `tile2data` is the data-space at level 0 (i.e. the highest resolution).

Slicing returns a chunk of image data. As with all image data we treat each pixel in that chunk as a sample. The `tile2data` transform tells us where those samples are in the data-space at level 0.

The other key constraint is that we want the sample at level 0 with the indices `(0, 0)` to be at the coordinate `(0.0, 0.0)` in that data space.

Therefore, in our current crude multi-scale approach, `tile2data` must take care of the offset and scale that take us from chunk sample indices to that level-0 data/index space.

For example, take the chunk index `(0, 1)`. According to the diagram, that needs to be transformed to `(0.5, 2.5)` in order to be consistent with the definitions above. This is achieved by a scale of `(2, 2)` and an offset of `(0.5, 0.5)`.

The other key factor here is that the vispy node texture needs another offset to account for the extent of the thing that it's rendering. That's what the code in `VispyBaseLayer._on_matrix_change` is trying to do.

Hope this helps. Will try to give this a bit more thought tomorrow.